library/core/src/mem/maybe_uninit.rs

```diff
@@ -1297,7 +1297,8 @@ impl<T, const N: usize> MaybeUninit<[T; N]> {
     /// let data: [MaybeUninit<u8>; 1000] = MaybeUninit::uninit().transpose();
     /// ```
     #[unstable(feature = "maybe_uninit_uninit_array_transpose", issue = "96097")]
-    pub fn transpose(self) -> [MaybeUninit<T>; N] {
+    #[inline(always)]
```

Comment:
If you make these just `#[inline]` instead of `#[inline(always)]`, I'll approve it right away.

But for `always` you need to make a case that it actually matters to have the stronger version.

Reply:
I guess I don't understand when `#[inline]` wouldn't fire vs. `#[inline(always)]`. As for why the stronger version matters, if the compiler ever decides not to inline we [end up with a memcpy](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:13,positionColumn:1,positionLineNumber:13,selectionStartColumn:1,selectionStartLineNumber:13,startColumn:1,startLineNumber:13),source:'use+core::mem::*%3B%0A%0Apub+fn+foo()+%7B%0A++++let+foo+%3D+transpose(%5BMaybeUninit::%3Cu8%3E::uninit()%3B+1024%5D)%3B%0A++++std::hint::black_box(%26foo)%3B%0A%7D%0A%0A%23%5Binline(never)%5D%0Apub+fn+transpose%3CT,+const+N:+usize%3E(this:+%5BMaybeUninit%3CT%3E%3B+N%5D)+-%3E+MaybeUninit%3C%5BT%3B+N%5D%3E+%7B%0A++++//+SAFETY:+T+and+MaybeUninit%3CT%3E+have+the+same+layout%0A++++unsafe+%7B+transmute_copy(%26ManuallyDrop::new(this))+%7D%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'--edition+2021+-C+opt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'rustc+nightly+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+nightly+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) whereas inlining [blocks this memcpy](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:13,positionColumn:1,positionLineNumber:13,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'use+core::mem::*%3B%0A%0Apub+fn+foo()+%7B%0A++++let+foo+%3D+transpose(%5BMaybeUninit::%3Cu8%3E::uninit()%3B+1024%5D)%3B%0A++++std::hint::black_box(%26foo)%3B%0A%7D%0A%0A%23%5Binline(always)%5D%0Apub+fn+transpose%3CT,+const+N:+usize%3E(this:+%5BMaybeUninit%3CT%3E%3B+N%5D)+-%3E+MaybeUninit%3C%5BT%3B+N%5D%3E+%7B%0A++++//+SAFETY:+T+and+MaybeUninit%3CT%3E+have+the+same+layout%0A++++unsafe+%7B+transmute_copy(%26ManuallyDrop::new(this))+%7D%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'--edition+2021+-C+opt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'rustc+nightly+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+nightly+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
Right, the inlining is clearly important.  But the main thing that, compared to nothing, `#[inline]` actually does  on *generic* methods is force it to be available in all [CGU](https://doc.rust-lang.org/rustc/codegen-options/index.html#codegen-units)s.  

Whereas `inline`-vs-`inline(always)` in comparison is just the difference between `inlinehint` and `alwaysinline` <https://llvm.org/docs/LangRef.html#function-attributes>, so we generally don't use `always`.  If LLVM really thinks that sometimes it's not worth doing despite the inlinehint, odds are that we should probably let it make that choice.  Sometimes we override it, where someone shows that it's worth it, but non-always `#[inline]` [seems sufficient](https://rust.godbolt.org/z/dh664rPhd) in your above example.

So let's start with non-always, and can always put it there later if needed.

Reply:
> Right, the inlining is clearly important. But the main thing that, compared to nothing, #[inline] actually does on generic methods is force it to be available in all [CGU](https://doc.rust-lang.org/rustc/codegen-options/index.html#codegen-units)s.

Interesting, was able to find [this](https://github.com/rust-lang/hashbrown/pull/119#issuecomment-537539046). I still don't think I get it, but it kind of seemed like inline leads to a potentially major hit and then always doesn't matter on top of that.

> If LLVM really thinks that sometimes it's not worth doing despite the inlinehint, odds are that we should probably let it make that choice.

Except that transpose is supposed to generate no code since it's just a cast. If LLVM doesn't inline, that would probably indicate a bug or some truly big-brain stuff. :laughing: 

Anyway, downgraded.